src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs

```diff
@@ -0,0 +1,476 @@
+
+                            return (CastResult)entryTargetAndResult;
+                        }
+                    }
+
+                    if (version == 0)
+                    {
+                        // the rest of the bucket is unclaimed, no point to search further
+                        break;
+                    }
+
+                    // quadratic reprobe
+                    i++;
+                    index = (index + i) & TableMask(table);
+                }
+            }
+            return CastResult.MaybeCast;
+        }
+
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        private static extern object IsInstanceOfAny_NoCacheLookup(void* toTypeHnd, object obj);
+
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        private static extern object ChkCastAny_NoCacheLookup(void* toTypeHnd, object obj);
+
+        // IsInstanceOf test used for unusual cases (naked type parameters, variant generic types)
+        // Unlike the IsInstanceOfInterface and IsInstanceOfClass functions,
+        // this test must deal with all kinds of type tests
+        [DebuggerHidden]
+        [StackTraceHidden]
+        [DebuggerStepThrough]
+        [MethodImpl(MethodImplOptions.AggressiveOptimization)]
+        private static object? IsInstanceOfAny(void* toTypeHnd, object? obj)
+        {
+            if (obj != null)
+            {
+                void* mt = RuntimeHelpers.GetMethodTable(obj);
+                if (mt != toTypeHnd)
+                {
+                    CastResult result = TryGet((nuint)mt, (nuint)toTypeHnd);
+                    if (result == CastResult.CanCast)
+                    {
+                        // do nothing
+                    }
+                    else if (result == CastResult.CannotCast)
+                    {
+                        obj = null;
+                    }
+                    else
+                    {
+                        goto slowPath;
+                    }
+                }
+            }
+
+            return obj;
+
+        slowPath:
+            // fall through to the slow helper
+            return IsInstanceOfAny_NoCacheLookup(toTypeHnd, obj);
+        }
+
+        [DebuggerHidden]
+        [StackTraceHidden]
+        [DebuggerStepThrough]
+        [MethodImpl(MethodImplOptions.AggressiveOptimization)]
+        private static object? IsInstanceOfInterface(void* toTypeHnd, object? obj)
+        {
+            if (obj != null)
+            {
+                MethodTable* mt = RuntimeHelpers.GetMethodTable(obj);
+                nuint interfaceCount = mt->InterfaceCount;
+                if (interfaceCount > 0)
+                {
+                    nuint* interfaceMap = mt->InterfaceMap;
+                    nuint i = 0;
+
+                    do
+                    {
+                        if (interfaceMap[i + 0] == (nuint)toTypeHnd)
+                            goto done;
+                        if (interfaceMap[i + 1] == (nuint)toTypeHnd)
```

Comment:
This overruns the buffer. Do we need to check that there is enough remaining slots?

Reply:
asm version did the same. The interface map is aligned to 4 entries and padded with zeros, if needed. That is so we could check in blocks of 4 entries. 
I will add a comment.

Reply:
Also added a comment about `ParentMethodTable`. There is some subtleness there too.

---
In reply to: [368308283](https://github.com/dotnet/runtime/pull/1068#discussion_r368308283) [](ancestors = 368308283)

Reply:
> asm version did the same

Can you add a link where it does it? I do not see it.

> The interface map is aligned to 4 entries and padded with zeroes, if needed. 

I do not think that we have any code that does alignment and padding like this. Also, I do not think it would be worth it - interface cast performance is not that important to pay for it.

Reply:
I think I got confused by comments in the asm version. 

---
In reply to: [368312453](https://github.com/dotnet/runtime/pull/1068#discussion_r368312453) [](ancestors = 368312453)

Reply:
Right, that comment is wrong.